VirtoCommerce.Platform.Web/Controllers/ExternalLoginController.cs

```diff
@@ -0,0 +1,200 @@
+        private readonly Func<ApplicationSignInManager> _signInManagerFactory;
+        private readonly Func<ApplicationUserManager> _userManagerFactory;
+        private readonly ISecurityService _securityService;
+        private readonly IEventPublisher _eventPublisher;
+        private readonly PlatformAuthenticationOptions _authenticationOptions;
+
+        public ExternalLoginController(Func<IAuthenticationManager> authenticationManagerFactory, Func<ApplicationSignInManager> signInManagerFactory,
+            Func<ApplicationUserManager> userManagerFactory, ISecurityService securityService, IEventPublisher eventPublisher,
+            PlatformAuthenticationOptions authenticationOptions)
+        {
+            _authenticationManagerFactory = authenticationManagerFactory;
+            _signInManagerFactory = signInManagerFactory;
+            _userManagerFactory = userManagerFactory;
+            _securityService = securityService;
+            _eventPublisher = eventPublisher;
+            _authenticationOptions = authenticationOptions;
+        }
+
+        [HttpGet]
+        [Route("externalsignin")]
+        [AllowAnonymous]
+        public ActionResult SignIn()
+        {
+            if (Request.IsAuthenticated)
+            {
+                return RedirectToAction("Index", "Home");
+            }
+
+            var returnUrl = Url.Action("Index", "Home");
+            var callbackUrl = Url.Action("SignInCallback", "ExternalLogin", new { returnUrl });
+
+            var authenticationProperties = new AuthenticationProperties { RedirectUri = callbackUrl };
+            authenticationProperties.Dictionary["LoginProvider"] = OpenIdConnectAuthenticationDefaults.AuthenticationType;
+
+            var authenticationManager = _authenticationManagerFactory();
+            authenticationManager.Challenge(authenticationProperties, OpenIdConnectAuthenticationDefaults.AuthenticationType);
+            return new EmptyResult();
+        }
+
+        [HttpGet]
+        [Route("externalsignin/callback")]
+        [AllowAnonymous]
+        public async Task<ActionResult> SignInCallback(string returnUrl)
+        {
+            var authenticationManager = _authenticationManagerFactory();
+            var externalLoginInfo = await authenticationManager.GetExternalLoginInfoAsync();
+            if (externalLoginInfo == null)
+            {
+                return Redirect(returnUrl);
+            }
+
+            var identity = externalLoginInfo.ExternalIdentity;
+            var userName = identity.FindFirstValue(ClaimTypes.Upn);
+            if (string.IsNullOrWhiteSpace(userName))
+            {
+                throw new InvalidOperationException("Received external login info does not have an UPN claim.");
+            }
+
+            var signInManager = _signInManagerFactory();
+            var externalLoginResult = await signInManager.ExternalSignInAsync(externalLoginInfo, false);
```

Comment:
There is some code duplications in `SignInExternalUser` and `CreateOrUpdatePlatformUser` methods
```
 await signInManager.SignInAsync(aspnetUser, true, true);
 await _eventPublisher.Publish(new UserLoginEvent(platformUser));
```

 I think it can be eliminated after some  little code refactoring to the next logic flow
1.  `ExternalSignInAsync` 
2. If exist user try to add external login and update user 
3. else need to create new user with external logins
4. user = userManager.FindByLoginInfo(loginInfo)
5. signInManager.SignIn(user)
6. publish event